### PR TITLE
Fix NPE in Stats Server Logging Handler

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Ensure that the stats server can handle requests to the `/logging` endpoint

# Context

## Gloo Edge Context
The logging endpoint is a useful feature for debugging components. In fact, we even recommend this feature in our production debugging docs for Gloo Edge: https://docs.solo.io/gloo-edge/latest/operations/debugging_gloo/#changing-logging-levels-and-more

## Problem
This feature has been broken for a while. You can see the [history](https://github.com/solo-io/go-utils/commits/master/stats/stats.go), but it has changed enough and been broken in different ways, that I decided to add a robust set of tests around the functionality.

In its most recent state, this handler is nil every time, unless the StartupOptions define a custom LogLevel.

## Solution
I introduced a context aspect to this server, to ensure that we can shutdown the server when the parent context is cancelled. Without this functionality, consecutive tests would fail, because the listener would never be shutdown and we would return stale config.

# How is this tested
- I added a variety of unit tests to ensure that we can use the logging endpoint to return the current log level, given a variety of startup options.
- I added a test to ensure that we can use the logging endpoint to change the log level, without requiring a restart to the server

I encountered some flakes, so to confirm that these tests don't introduce any flakes, I ran:
```
ginkgo -r -failFast -trace -progress -compilers=4 -failOnPending -noColor -randomizeSuites   -randomizeAllSpecs -untilItFails stats
```
and let it run for a while, before quitting

### CODE REVIEW

File: stats/stats.go
Comment (elcasteel) [inline stats/stats.go:83]: Why did you remove this `else`?

File: stats/stats.go
Comment (ben-taussig-solo) [inline stats/stats.go:71]: You touched on it a little in slack, but could you clarify why we're moving this mux setup out of the goroutine?

File: stats/stats.go
Comment (elcasteel) [reply stats/stats.go:83]: Nevermind I figured out where it moved

Comment (sam-heilbron) [reply stats/stats.go:83]: Yeah I found the code a little confusing because we set the level in some places and get it from others. To try to simplify, I tried to separate those two

File: stats/stats.go
Comment (sam-heilbron) [reply stats/stats.go:71]: I wanted to be able to have another goroutine that could shutdown the server. I ran into the issue when writing tests that a single test would pass, but running multiple would fail, because the original goroutine was leaking / still running.

i'm going to move this construction back to the goroutine, back to how it used to be

**Review by ben-taussig-solo — COMMENTED** [review]
Looks good to me. Waiting on tests + Ethan's review before approval.

File: stats/stats.go
Comment (jackstine) [inline stats/stats.go:104]: LGTM, just saw that this was changed.

**Comment by jackstine** [comment]
LGTM, I'll let Ethan approve it

File: stats/stats.go
Comment (kdorosh) [inline stats/stats.go:94]: unhandled error

File: stats/stats.go
Comment (kdorosh) [inline stats/stats.go:101]: unhandled error. at least log?

File: stats/stats.go
Comment (sam-heilbron) [reply stats/stats.go:94]: Good call! It wasn't before and I lazily (incorrectly) just maintained. Added in a log line that should make things clearer

File: stats/stats_suite_test.go
Comment (kdorosh) [inline stats/stats_suite_test.go:30]: nit, prefer `os.Unsetenv`

File: stats/stats_test.go
Comment (kdorosh) [inline stats/stats_test.go:47]: nit, prefer `os.Unsetenv()`

File: stats/stats.go
Comment (kdorosh) [inline stats/stats.go:74]: this leaks a goroutine itself, if we cancel the context this goroutine does not terminate

File: stats/stats.go
Comment (kdorosh) [inline stats/stats.go:81]: this seems like a race too.. what if the caller depended on the `addhandlers` array behind the slice copy passed in? I think we should copy the array contents before using them in the goroutine

File: stats/stats.go
Comment (kdorosh) [inline stats/stats.go:106]: nit, no need to select with a single case. we can just do `<-ctx.Done()`
```suggestion
		<-ctx.Done()
````

**Review by kdorosh — COMMENTED** [review]
lgtm, leaving for others to approve

File: stats/stats.go
Comment (sam-heilbron) [reply stats/stats.go]: great catch here, updated

File: stats/stats.go
Comment (sam-heilbron) [reply stats/stats.go]: yeah good call, updated

**Review by elcasteel — APPROVED** [review]

File: stats/stats_test.go
Comment (sam-heilbron) [reply stats/stats_test.go:59]: There should only be 1 beforeEach per test that call this. The reason for doing it was that each set of tests require different startup options to run with. I figured my options were:
1. Define the startupOptions in the BeforeEach of each test group, and call StartStatsServer in a JustBeforeEach at the outer Context
2. Call StartStatsServer in each of the individual `It` and letting the startupOptions be constructed in the BeforeEach
3. This approach


I'm not strongly bound to this and am certainly open to an alternative approach if you have one in mind that you think would be easier to reason through

Comment (elcasteel) [reply stats/stats_test.go:59]: Ah, the narrow github diff makes reading indentation levels harder and I was _convinced_ that the second call was nested in this first one. 
This approach seems fine to me. I don't think there's a a great way to handle the way ginkgo tests end up very deeply nested.

**Pull request merged**
